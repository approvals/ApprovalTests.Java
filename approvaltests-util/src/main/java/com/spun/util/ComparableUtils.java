package com.spun.util;

public class ComparableUtils
{
  public static <T extends Comparable<T>> boolean isAEqualToB(T a, T b)
  {
    return a.compareTo(b) == 0;
  }

  public static <T extends Comparable<T>> boolean isALessThanOrEqualToB(T a, T b)
  {
    return a.compareTo(b) <= 0;
  }

  public static <T extends Comparable<T>> boolean isALessThanB(T a, T b)
  {
    return a.compareTo(b) < 0;
  }

  public static <T extends Comparable<T>> ComparableWrapper<T> wrap(T object)
  {
    return new ComparableWrapper<>(object);
  }
  public static class ComparableWrapper<T extends Comparable<T>>
  {
    private final T object;
    public ComparableWrapper(T object)
    {
      this.object = object;
    }

    public boolean isLessThan(T other)
    {
      return isALessThanB(object, other);
    }

    public boolean isEqualTo(T other)
    {
      return isAEqualToB(object, other);
    }

    public boolean isLessThanOrEqual(T other)
    {
      return isALessThanOrEqualToB(object, other);
    }
  }
}
