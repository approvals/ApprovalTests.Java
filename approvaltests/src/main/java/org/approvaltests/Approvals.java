package org.approvaltests;

import com.spun.util.ArrayUtils;
import com.spun.util.FormattedException;
import com.spun.util.JsonUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.persistence.ExecutableCommand;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.*;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.MasterDirectoryNamer;
import org.approvaltests.namer.NamerFactoryForOptions;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class Approvals
{
  public static final NamerFactoryForOptions NAMES        = NamerFactoryForOptions.INSTANCE;
  public static Loader<ApprovalNamer>        namerCreater = new Loader<ApprovalNamer>()
                                                          {
                                                            public ApprovalNamer load()
                                                            {
                                                              return new StackTraceNamer();
                                                            }
                                                          };
  public static void verify(String response)
  {
    verify(response, new Options());
  }
  public static void verify(String response, Options options)
  {
    verify(options.createWriter(response), options);
  }
  public static void verify(Object object)
  {
    verify(object, new Options());
  }
  public static void verify(Object object, Options options)
  {
    checkForAwtComponents(object == null ? Object.class : object.getClass());
    verify(Objects.toString(object), options);
  }
  public static void verify(Verifiable object)
  {
    verify(object, new Options());
  }
  public static void verify(Verifiable object, Options options)
  {
    VerifyParameters verifyParameters = object.getVerifyParameters(options);
    verify(Objects.toString(object), verifyParameters.getOptions());
  }
  private static void checkForAwtComponents(Class<?> type)
  {
    if (type == Object.class)
    { return; }
    if ("java.awt.Component".equals(type.getCanonicalName()))
    {
      throw new FormattedException(
          "Approvals.verify(Component) has moved to AwtApprovals.verify(Component)\ncalled Approvals.verify(Object)");
    }
    checkForAwtComponents(type.getSuperclass());
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, T[] array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, T[] array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1, Options options)
  {
    verifyAll("", values, f1, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1, Options options)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(text, options);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1, new Options());
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1, Options options)
  {
    verifyAll(header, Arrays.asList(values), f1, options);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    verifyAll(header, array, f1, new Options());
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1, Options options)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(text, options);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, Iterable<T> array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static void verifyHtml(String response)
  {
    verifyHtml(response, new Options());
  }
  public static void verifyHtml(String response, Options options)
  {
    verify(response, options.forFile().withExtension(".html"));
  }
  public static void verify(File generateFile)
  {
    verify(generateFile, new Options());
  }
  public static void verify(File generateFile, Options options)
  {
    verify(options.createWriter(generateFile), options);
  }
  /**
   * @deprecated Use {@link #verify(ApprovalWriter, ApprovalNamer, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), new Options(reporter));
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer)
  {
    verify(writer, namer, new Options());
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, Options options)
  {
    verify(new FileApprover(writer, namer, options.getComparator()), options);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, new Options());
  }
  public static void verify(ApprovalWriter writer, Options options)
  {
    verify(writer, options.forFile().getNamer(), options);
  }
  public static void verifyXml(String xml)
  {
    verifyXml(xml, new Options());
  }
  public static void verifyXml(String xml, Options options)
  {
    final String formattedXml = ApprovalXmlWriter.prettyPrint(xml, 2);
    verify(formattedXml, options.forFile().withExtension(".xml"));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, new Options());
  }
  /**
   * @deprecated Use {@link #verify(ApprovalApprover, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, new Options(reporter));
  }
  public static void verify(ApprovalApprover approver, Options options)
  {
    ApprovalFailureReporter reporter = options.getReporter();
    VerifyResult result = approver.approve();
    if (result.isFailure())
    {
      result = approver.reportFailure(reporter);
    }
    if (result.isSuccessful())
    {
      approver.cleanUpAfterSuccess(reporter);
    }
    else
    {
      approver.fail();
    }
  }
  public static void verify(ExecutableCommand query)
  {
    verify(query, new Options());
  }
  public static void verify(ExecutableCommand query, Options options)
  {
    verify(query.getCommand(), ExecutableQueryFailure.create(query, options));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(map, new Options());
  }
  public static void verify(Map<?, ?> map, Options options)
  {
    verify(StringUtils.toString(map), options);
  }
  public static void verify(ResultSet rs)
  {
    verify(rs, new Options());
  }
  public static void verify(ResultSet rs, Options options)
  {
    verify(options.createWriter(rs), options);
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(loader, new Options());
  }
  public static <T> void verify(SqlLoader<T> loader, Options options)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader), options);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileInDirectory(directory, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(), options);
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files, Options options)
  {
    List<File> mismatched = new ArrayList<>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(options.createWriter(f), new MasterDirectoryNamer(f, options), options);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json)
  {
    JsonApprovals.verifyJson(json);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json, Options options)
  {
    JsonApprovals.verifyJson(json, options);
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(actual, fileExtensionWithoutDot, new Options());
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot, Options options)
  {
    verify("" + actual, options.forFile().withExtension(fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o)
  {
    JsonApprovals.verifyAsJson(o);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass)
  {
    verifyAsJson(o, gsonBuilder, gsonBuilderClass, new Options());
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass,
      Options options)
  {
    verify(JsonUtils.asJsonWithBuilder(o, gsonBuilder, gsonBuilderClass),
        options.forFile().withExtension(".json"));
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o, Options options)
  {
    JsonApprovals.verifyAsJson(o, options);
  }
  public static void verifyException(Action0 runnableBlock)
  {
    verifyException(runnableBlock, new Options());
  }
  public static void verifyException(Action0 runnableBlock, Options options)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null)
    { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()), options);
  }
  public static ApprovalSettings settings()
  {
    return new ApprovalSettings();
  }
}
