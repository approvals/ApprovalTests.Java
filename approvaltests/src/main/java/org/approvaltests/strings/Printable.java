package org.approvaltests.strings;

import org.lambda.functions.Function1;

import java.util.ArrayList;
import java.util.Arrays;

public class Printable<T>
{
  private T                          value;
  private final Function1<T, String> toString;
  public Printable(T value, Function1<T, String> toString)
  {
    this.value = value;
    this.toString = toString;
  }

  public Printable(T value, String label)
  {
    this(value, __ -> label);
  }

  public static <T> Printable<T>[] create(Function1<T, String> toString, T... parameters)
  {
    return Arrays.stream(parameters).map(p -> new Printable<>(p, toString)).toArray(Printable[]::new);
  }

  public static LabelMakerStarter with()
  {
    return new LabelMakerStarter();
  }

  public T get()
  {
    return value;
  }

  @Override
  public String toString()
  {
    return toString.call(value);
  }
  public static class LabelMakerStarter
  {
    public <T> LabelMaker<T> label(T value, String label)
    {
      LabelMaker<T> maker = new LabelMaker<T>();
      return maker.label(value, label);
    }
  }
  public static class LabelMaker<T>
  {
    private ArrayList<Printable<T>> values = new ArrayList<>();
    public LabelMaker<T> label(T value, String label)
    {
      values.add(new Printable(value, label));
      return this;
    }

    public Printable<T>[] toArray()
    {
      return values.toArray(new Printable[0]);
    }
  }
}
