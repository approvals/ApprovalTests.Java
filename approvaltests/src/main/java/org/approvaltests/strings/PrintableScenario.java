package org.approvaltests.strings;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.Callable;

public class PrintableScenario {


    private final StringBuilder toVerify = new StringBuilder();
    private final String name;
    private final String description;
    private final ArrayList<Printable> printables = new ArrayList<>();

    public PrintableScenario(String name, String description) {
        this.name = name;
        this.description = description;
    }
    public PrintableScenario(String name) {
        this(name, "");
    }

    public void given(Printable... printables) {
        arrange(printables);
    }

    public void arrange(Printable... printables) {
        this.printables.addAll(Arrays.asList(printables));
        toVerify.append(makeHeading(name, description, "="));
        toVerify.append("\n");
        toVerify.append(printAll());
    }

    public String makeHeading(String heading, String summary, String underlineCharacter) {
        int count = heading.length();
        // create a string made up of n copies of string "="
        String underlineHeading = String.join("", Collections.nCopies(count, underlineCharacter));
        return heading + "\n" + underlineHeading + "\n" + summary + "\n";
    }

    public String printAll() {
        StringBuilder result = new StringBuilder();
        for (Printable printable : this.printables) {
            result.append(printable.toString());
            result.append("\n");
        }
        return result.toString();
    }

    public void when(String action, Callable function) throws Exception {
        function.call();
        when(action);
    }

    public void when(String action) {
        act(action);
    }

    public void act(String action) {
        toVerify.append(makeHeading(action, "", "-"));
        toVerify.append(printAll());
    }
    public String then() {
        return print();
    }

    public String print() {
        return toVerify.toString();
    }
}
