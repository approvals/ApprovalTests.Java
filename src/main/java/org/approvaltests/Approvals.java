package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.fs.Path;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalBinaryFileWriter;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.DirectoryToDirectoryWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.ResultSetApprovalWriter;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.lambda.functions.Function1;
import org.lambda.functions.implementations.F1;
import org.lambda.query.Query;

import com.spun.util.ArrayUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.images.ImageWriter;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.SqlLoader;

public class Approvals
{
  /**
   * @deprecated Use verify()
   */
  public static void approve(String response) throws Exception
  {
    verify(response);
  }
  public static void verify(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
  }
  public static void verify(Object o) throws Exception
  {
    verify("" + o);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, T[] array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, T[] array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, Iterable<T> array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, Iterable<T> array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Component c) throws Exception
  {
    verify(c);
  }
  public static void verify(Component c) throws Exception
  {
    verify(new ComponentApprovalWriter(c), FileTypes.Image);
  }
  /**
   * @deprecated Use verifyHtml()
   */
  public static void approveHtml(String response) throws Exception
  {
    verifyHtml(response);
  }
  public static void verifyHtml(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(File generateFile)
  {
    verify(generateFile);
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile), FileTypes.File);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Image image)
  {
    verify(image);
  }
  public static void verify(Image image)
  {
    approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(BufferedImage bufferedImage)
  {
    verify(bufferedImage);
  }
  public static void verify(BufferedImage bufferedImage)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(writter, namer, reporter);
  }
  public static void verify(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writter, namer), reporter);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, String fileType)
  {
    verify(writter, fileType);
  }
  public static void verify(ApprovalWriter writter, String fileType)
  {
    verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
  }
  /**
   * @deprecated Use verifyXml()
   */
  public static void approveXml(String xml) throws Exception
  {
    verifyXml(xml);
  }
  public static void verifyXml(String xml) throws Exception
  {
    verify(new ApprovalXmlWriter(xml), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(FileApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, reporter);
  }
  public static void verify(FileApprover approver, ApprovalFailureReporter reporter)
  {
    try
    {
      if (!approver.approve())
      {
        boolean passed = false;
        if (reporter instanceof ApprovalFailureOverrider)
        {
          passed = approver.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
        }
        if (!passed)
        {
          approver.reportFailure(reporter);
          approver.fail();
        }
        else
        {
          approver.cleanUpAfterSuccess(reporter);
        }
      }
      else
      {
        approver.cleanUpAfterSuccess(reporter);
      }
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ExecutableQuery query) throws Exception
  {
    verify(query);
  }
  public static void verify(ExecutableQuery query) throws Exception
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), new JUnitStackTraceNamer(),
        new ExecutableQueryFailure(query));
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Map map) throws Exception
  {
    verify(map);
  }
  public static void verify(Map map) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(RackResponse response) throws Exception
  {
    verify(response);
  }
  public static void verify(RackResponse response) throws Exception
  {
    if (isImage(response))
    {
      String fileType = "png";
      verify(new ApprovalBinaryFileWriter(response.getResponse(), fileType), fileType);
    }
    else
    {
      verifyHtml(response.getResponse().toString());
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ResultSet rs) throws Exception
  {
    verify(rs);
  }
  public static void verify(ResultSet rs) throws Exception
  {
    verify(new ResultSetApprovalWriter(rs), "csv");
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(SqlLoader loader) throws Exception
  {
    verify(loader);
  }
  public static void verify(SqlLoader loader) throws Exception
  {
    verify(new SqlLoader.ExecutableWrapper(loader));
  }
  private static boolean isImage(RackResponse response)
  {
    String type = response.getHeaders().get(RackResponseUtils.CONTENT_TYPE);
    return RackResponseUtils.CONTENT_TYPE_IMAGE.equals(type);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return new JUnitStackTraceNamer();
  }
  private static void approve(BufferedImage bufferedImage, ApprovalNamer namer)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + Path.SEPARATOR + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory), FileTypes.File);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, new F1<File, String>(mismatched.get(1))
    {
      {
        ret(a.getName());
      }
    }).toString();
  }
}