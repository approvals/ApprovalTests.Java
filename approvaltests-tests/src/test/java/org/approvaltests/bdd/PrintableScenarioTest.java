package org.approvaltests.bdd;

import org.approvaltests.Approvals;
import org.approvaltests.strings.Printable;
import org.approvaltests.strings.PrintableScenario;
import org.junit.Test;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class PrintableScenarioTest
{
  @Test
  public void printBDDScenario() throws Exception
  {
    PrintableScenario story = new PrintableScenario("Buy Groceries", "Bob puts two items in the shopping cart");
    User currentUser = new User();
    ShoppingCart shoppingCart = new ShoppingCart(currentUser);
    story.given(new Printable<>(currentUser, UserPrinter::print),
        new Printable<>(shoppingCart, ShoppingCartPrinter::print));
    story.when("Add oranges to the cart", () -> {
      shoppingCart.add("Oranges", 1, BigDecimal.valueOf(5));
      return null;
    });
    story.when("Add apples to the cart", () -> {
      shoppingCart.add("Apples", 3, BigDecimal.valueOf(4));
      return null;
    });
    Approvals.verify(story.then());
  }
}

class User
{
  public String name = "Bob";
}

class Item
{
  public String     name;
  public Integer    quantity;
  public BigDecimal price;
  public BigDecimal amount()
  {
    return price.multiply(BigDecimal.valueOf(quantity));
  }
  public Item(String itemName, int quantity, BigDecimal price)
  {
    this.name = itemName;
    this.quantity = quantity;
    this.price = price;
  }
}

class ShoppingCart
{
  private final User user;
  public List<Item>  articles = new ArrayList<>();
  public ShoppingCart(User currentUser)
  {
    this.user = currentUser;
  }
  public BigDecimal subtotal()
  {
    return articles.stream().map(Item::amount).reduce(BigDecimal.ZERO, BigDecimal::add);
  }
  public BigDecimal shipping()
  {
    if (Objects.equals(user.name, "Bob"))
    { return BigDecimal.ONE; }
    return BigDecimal.ZERO;
  }
  public BigDecimal total()
  {
    return this.subtotal().add(this.shipping());
  }
  public void add(String itemName, int quantity, BigDecimal price)
  {
    this.articles.add(new Item(itemName, quantity, price));
  }
}

class ShoppingCartPrinter
{
  public static String print(ShoppingCart shoppingCart)
  {
    StringBuilder result = new StringBuilder();
    result.append("ShoppingCart:\n");
    result.append("Articles:\n");
    result.append(shoppingCart.articles.stream()
        .map(article -> "    " + article.name + " price: " + article.price + " quantity: " + article.quantity)
        .collect(Collectors.joining("\n")));
    result.append("\n");
    result.append("Subtotal:" + shoppingCart.subtotal() + "\n");
    result.append("Shipping:" + shoppingCart.shipping() + "\n");
    result.append("Total:" + shoppingCart.total() + "\n");
    return result.toString();
  }
}

class UserPrinter
{
  public static String print(User user)
  {
    return "User: " + user.name + "\n";
  }
}