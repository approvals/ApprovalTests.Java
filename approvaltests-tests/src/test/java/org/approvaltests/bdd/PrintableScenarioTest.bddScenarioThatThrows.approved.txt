Shoplifting
===========
We expect an exception

User: Bob

ShoppingCart:
Articles:

Subtotal:0
Shipping:1
Total:1

Something illegal
-----------------
This action threw an exception: java.lang.RuntimeException: shoplifting

User: Bob

ShoppingCart:
Articles:

Subtotal:0
Shipping:1
Total:1

